docs/proposals/01-extensibility.md

```diff
@@ -0,0 +1,844 @@
+* Gardener does not contain any operation system-specific knowledge anymore but defines a clear contract allowing external controllers to support different operation systems (CoreOS, SLES, Ubuntu, ...).
+* It shall become much easier to move control planes of Shoot clusters between Seed clusters ([#232](https://github.com/gardener/gardener/issues/232)) which is a necessary requirement of an automated setup for the Gardener Ring ([#233](https://github.com/gardener/gardener/issues/233)).
+
+### Non-Goals
+
+* We want to also factor out the specific knowledge of the addon deployments (nginx-ingress, kubernetes-dashboard, ...), but we already have dedicated projects/issues for that: https://github.com/gardener/bouquet and [#246](https://github.com/gardener/gardener/issues/246). We will keep the addons in-tree as part of this proposal and tackle their extraction separately.
+* We do not want to make the Gardener a plain workflow engine that just executes a given template (which indeed would allow to be generic, open, and extensible in their highest forms but which would end-up in building a "programming/scripting language" inside a serialization format (YAML/JSON/...)). Rather, we want to have well-defined contracts and APIs.
+
+## Proposal
+
+Gardener heavily relies on and implements Kubernetes principles, and its first goal is to use Kubernetes wherever applicable.
+The extension concept in Kubernetes embraces (next to others) `CustomResourceDefinition`s as well as `ValidatingWebhookConfiguration`s and `MutatingWebhookConfiguration`s.
+Consequently, Gardener's extensibility concept relies on these mechanisms.
+
+Instead of doing everything on its own, Gardener will deploy a bunch of CRDs to the Seed cluster which will be watched by dedicated controllers only containing the specific knowledge of their function area.
+Gardener will just wait until the controllers report that they are done (or have faced an error) in the CRD's `.status` field instead of doing the respective tasks itself.
+We will have one specific CRD for every specific operation (e.g., DNS, infrastructure provisioning, machine cloud config generation, ...).
+However, there are also parts inside Gardener which are handled generally (not by cloud botanists) because they are the same or very similar for all the environments.
+One example of those is the deployment of a `Namespace` in the Seed which will run the Shoot's control plane
+Another one is the deployment of a `Service` for the Shoot's kube-apiserver.
+In case a cloud botanist needs to interfere and react on those operations it should register a `ValidatingWebhookConfiguration` or a `MutatingWebhookConfiguration`.
+With this approach it can validate or modify any resource created by Gardener to make it cloud specific.
+
+### Modification of existing `CloudProfile` and `Shoot` resources
+
+We will introduce the new API group `gardener.cloud`:
```

Comment:
I struggle to imagine how this would be handled during the transition. Some very simple questions pop in my mind e.g.: how would i query old and new CloudProvider resources if at a point they coexist.

The plan could be creating a new major release and provide some methods, scripts to handle migration, but shouldn't this proposal clarify this process as well? (I'm not sure if it should or not)

Reply:
At the moment, I imagine the conversion to happen automatically. Do you think that's a problem? `Deployment`s were also moved from `extensions` to `apps` API group, we could do the same.
If we say it's too complicated we can follow @bergerx' approach to create a major release with proper migration docs. But I guess the automatic conversion is feasible?